CC Crash `@retains[?]` in munit-style code at same time as compiling sealed ADT

## Compiler version

3.8.4-RC1


## Minimized code

`no TypeBounds allowed` in pickler with pickling some `@retains[?]` annotation that i guess shouldn't be there?

<!--
This code should be self contained, compilable (with possible failures) and as small as possible.

Ideally, we should be able to just copy this code to a file and run `scalac` (and maybe `scala`) to reproduce the issue.

If the code has external dependencies, please provide the Scala CLI directives (or SBT/other build tool configuration) that describe them.
Also note that it's easier and faster for the maintenance team to address issues minimised to reproduce bugs without external dependencies.

It's most convenient to also include `using` directives for the Scala version that demonstrates the problem,
any compiler command-line options, as well as dependencies. An example is provided.

It's also fine to paste the transcript of a REPL session. Note that some bugs may be specific to the REPL.
-->

```scala
//> using scala 3.8.4-RC1
import language.experimental.captureChecking

sealed trait Res[+T]:
  def toOption: Option[T] = this match
    case Ok(value) => Some(value)
    case _         => None

case class Ok[+T](value: T) extends Res[T]
case object Bad extends Res[Nothing]

def check[A](obtained: A, clue: => Any = "not equal"): Unit =
  assert(obtained != null, clue)

def test(): Unit =
  val evens = (1 to 5).map(v => if v % 2 == 0 then Ok(v) else Bad).flatMap(_.toOption)
  check(evens)
```

## Output

```scala
Error: java.lang.AssertionError: no TypeBounds allowed
        at dotty.tools.dotc.core.TypeApplications$.noBounds(TypeApplications.scala:23)
        at dotty.tools.dotc.core.TypeApplications$.argTypes$extension$$anonfun$1(TypeApplications.scala:565)
        at scala.collection.immutable.List.loop$3(List.scala:466)
        at scala.collection.immutable.List.mapConserve(List.scala:491)
        at dotty.tools.dotc.core.TypeApplications$.argTypes$extension(TypeApplications.scala:565)
        at dotty.tools.dotc.ast.tpd$.New(tpd.scala:577)
        at dotty.tools.dotc.ast.tpd$.New(tpd.scala:573)
        at dotty.tools.dotc.core.Annotations$CompactAnnotation.oldTree(Annotations.scala:146)
        at dotty.tools.dotc.core.tasty.TreePickler.pickleNewType(TreePickler.scala:294)
        at dotty.tools.dotc.core.tasty.TreePickler.pickleType(TreePickler.scala:184)
        at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:729)
        at dotty.tools.dotc.core.tasty.TreePickler.pickleTpt(TreePickler.scala:364)
        at dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$4(TreePickler.scala:546)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
        at scala.collection.immutable.List.foreach(List.scala:327)
        at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:546)
        at dotty.tools.dotc.core.tasty.TreePickler.pickleTree$$anonfun$3(TreePickler.scala:539)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
        at scala.collection.immutable.List.foreach(List.scala:327)
        at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:539)
        at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:583)
        at dotty.tools.dotc.core.tasty.TreePickler.pickleTreeUnlessEmpty(TreePickler.scala:368)
        at dotty.tools.dotc.core.tasty.TreePickler.pickleDef(TreePickler.scala:402)
        at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:675)
        at dotty.tools.dotc.core.tasty.TreePickler.pickleStats$$anonfun$2(TreePickler.scala:435)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
        at scala.collection.immutable.List.foreach(List.scala:327)
        at dotty.tools.dotc.core.tasty.TreePickler.pickleStats(TreePickler.scala:435)
        at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:711)
        at dotty.tools.dotc.core.tasty.TreePickler.pickleDef(TreePickler.scala:393)
        at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:677)
        at dotty.tools.dotc.core.tasty.TreePickler.pickleStats$$anonfun$2(TreePickler.scala:435)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
        at scala.collection.immutable.List.foreach(List.scala:327)
        at dotty.tools.dotc.core.tasty.TreePickler.pickleStats(TreePickler.scala:435)
        at dotty.tools.dotc.core.tasty.TreePickler.pickleTree(TreePickler.scala:727)
        at dotty.tools.dotc.core.tasty.TreePickler.pickle$$anonfun$1(TreePickler.scala:954)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
        at scala.collection.immutable.List.foreach(List.scala:327)
        at dotty.tools.dotc.core.tasty.TreePickler.pickle(TreePickler.scala:952)
        at dotty.tools.dotc.transform.Pickler.run$$anonfun$1$$anonfun$1(Pickler.scala:306)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
        at scala.collection.immutable.List.foreach(List.scala:327)
        at dotty.tools.dotc.transform.Pickler.run$$anonfun$1(Pickler.scala:279)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
        at scala.collection.immutable.List.foreach(List.scala:327)
        at dotty.tools.dotc.transform.Pickler.run(Pickler.scala:278)
        at dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:411)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
        at scala.collection.immutable.List.foreach(List.scala:327)
        at dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:403)
        at dotty.tools.dotc.transform.Pickler.runPhase$1(Pickler.scala:392)
        at dotty.tools.dotc.transform.Pickler.runOn(Pickler.scala:398)
        at dotty.tools.dotc.Run.runPhases$1$$anonfun$1(Run.scala:380)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
        at scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1324)
        at dotty.tools.dotc.Run.runPhases$1(Run.scala:373)
        at dotty.tools.dotc.Run.compileUnits$$anonfun$1$$anonfun$2(Run.scala:420)
        at dotty.tools.dotc.Run.compileUnits$$anonfun$1$$anonfun$adapted$1(Run.scala:420)
        at scala.Function0.apply$mcV$sp(Function0.scala:42)
        at dotty.tools.dotc.Run.showProgress(Run.scala:482)
        at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:420)
        at dotty.tools.dotc.Run.compileUnits$$anonfun$adapted$1(Run.scala:432)
        at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:69)
        at dotty.tools.dotc.Run.compileUnits(Run.scala:432)
        at dotty.tools.dotc.Run.compileSources(Run.scala:319)
        at dotty.tools.dotc.Run.compile(Run.scala:304)
        at dotty.tools.dotc.Driver.doCompile(Driver.scala:37)
```

## Expectation

--- a/compiler/src/dotty/tools/dotc/typer/Applications.scala
+++ b/compiler/src/dotty/tools/dotc/typer/Applications.scala
@@ -486,7 +486,7 @@ object Applications {
       // it's crucial that the type tree is not copied directly as argument to
       // `cpy$default$1`. If it was, the variable `X'` would already be interpolated
       // when typing the default argument, which is too early.
-      spliceMeth(meth, fn).appliedToTypeTrees(targs.map(targ => TypeTree(targ.tpe).withSpan(targ.span)))
+      spliceMeth(meth, fn).appliedToTypeTrees(targs.map(targ => TypeTree(targ.tpe, inferred = true).withSpan(targ.span)))
     case _ => meth
   }
 

--- a/tests/pos-custom-args/captures/i25849.scala
+++ b/tests/pos-custom-args/captures/i25849.scala
@@ -1,0 +1,16 @@
+import language.experimental.captureChecking
+
+sealed trait Res[+T]:
+  def toOption: Option[T] = this match
+    case Ok(value) => Some(value)
+    case _         => None
+
+case class Ok[+T](value: T) extends Res[T]
+case object Bad extends Res[Nothing]
+
+def check[A](obtained: A, clue: => Any = "not equal"): Unit =
+  assert(obtained != null, clue)
+
+def test(): Unit =
+  val evens = (1 to 5).map(v => if v % 2 == 0 then Ok(v) else Bad).flatMap(_.toOption)
+  check(evens)

